### PR TITLE
Auth lists tenants and projects only current user data

### PR DESCRIPTION
**Pull request by savelyev-an**
1) Some refactoring folder structure of flant_iam_auth to unify with flant_iam
2) Implement filtering tenants and projects at flant_iam_auth

### CODE REVIEW

File: vault-plugins/flant_iam_auth/backend/backend.go [hunk 1]
```diff
@@ -348,3 +351,196 @@ const (
...
+			defer txn.Abort()
+			groups, err := iam_repo.NewGroupRepository(txn).FindAllParentGroupsForUserUUID(user.UUID)
+			gs := make([]iam.GroupUUID, 0, len(groups))
+			for uuid := range groups {
+				gs = append(gs, uuid)
+			}
+			if err != nil {
+				return nil, nil, fmt.Errorf("collecting tenants, get groups: %w", err)
+			}
+			//  TODO Here easy to have two or three paralleled goroutines
```
Comment (shvgn) [inline vault-plugins/flant_iam_auth/backend/backend.go:441]: I'd say it's arguable whether concurrency is worthwhile here

File: vault-plugins/flant_iam_auth/backend/path_multipass_owner.go [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+				err := fmt.Errorf("can't cast, need *model.User, got: %T", subject)
+				logger.Debug(err.Error())
+				return responseErrMessage(req, err.Error(), http.StatusInternalServerError)
+			}
+			return logical.RespondWithStatusCode(&logical.Response{
+				Data: map[string]interface{}{
+					"user": model.User{
+						UUID:             user.UUID,
+						TenantUUID:       user.TenantUUID,
+						Origin:           user.TenantUUID,
```
[the remark this answers is not in the document]
Comment (shvgn) [reply vault-plugins/flant_iam_auth/backend/path_multipass_owner.go:61]: I observed tenant ID used as origin in other places, not related this PR

File: vault-plugins/flant_iam_auth/backend/backend.go [hunk 3]
```diff
@@ -348,3 +351,196 @@ const (
...
+			if err != nil {
+				return nil, nil, fmt.Errorf("collecting projects: %w", err)
+			}
+			return tenants, projects, nil
+		}
+	}
+	return nil, nil, fmt.Errorf("unexpected subjectType: `%s`", subjectType)
+}
+
+func collectProjectUUIDsFromRoleBindigns(rbs1 map[iam.RoleBindingUUID]*iam.RoleBinding,
```
Comment (shvgn) [inline vault-plugins/flant_iam_auth/backend/backend.go:506]: typo: collectProjectUUIDsFromRoleBindi**ng**s

File: vault-plugins/flant_iam_auth/backend/path_tenant.go [hunk 4]
```diff
@@ -129,15 +135,25 @@ func (b *flantIamAuthBackend) listProjects(ctx context.Context, req *logical.Req
...
+		if tenantAcccepted && projectAcccepted {
+			res := model.SafeProject{
+				UUID:       project.UUID,
+				TenantUUID: project.TenantUUID,
+				Version:    project.Version,
+			}
+			result = append(result, res)
 		}
-		result = append(result, res)
 	}
```
Comment (shvgn) [inline vault-plugins/flant_iam_auth/backend/path_tenant.go:157]: I'd say the code is somewhat messy here. The vault framework detail dives too deeply. This transport layer does too much from its behalf:

1. takes tx
2. calls availableTenantsAndProjectsByEntityIDOwner(ctx, req)
  3. also takes tx
  4. calls revealEntityIDOwner(ctx, req)
    5. takes tx 
    6. context is not used
    7. req is not used, only EntityID is of use
    6. calls IdentityAPI which is not injected

I would expect lines 133–156 to be a usecase one-line call. NewIdentityAPI could be instantiated here and passed to the usecase as dependency

I would suggest making one more iteration on that

**Review by shvgn — COMMENTED** [review]
I would not request changes in terms ofrefactoring, but I would suggest them.

File: vault-plugins/flant_iam_auth/backend/path_multipass_owner.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (savelyev-an) [reply vault-plugins/flant_iam_auth/backend/path_multipass_owner.go:61]: fixed

File: vault-plugins/flant_iam/repo/rolebinding.go [hunk 5]
```diff
@@ -359,61 +300,57 @@ func (r *RoleBindingRepository) FindDirectRoleBindingsForTenantGroups(tenantUUID
...
+	iter, err := r.db.Get(model.RoleBindingType, ProjectInRoleBindingIndex, projectUUID)
 	if err != nil {
 		return nil, err
 	}
-	return extractRoleBindings(iter)
+	return extractRoleBindings(iter, false)
 }
 
-type roleInTenantRoleBindingIndexer struct{}
+// roleInRoleBindingIndexer build index rb.Roles[i].Name, several indexes for one record
```
[the remark this answers is not in the document]
Comment (savelyev-an) [reply vault-plugins/flant_iam/repo/rolebinding.go:312]: fixed

File: vault-plugins/flant_iam_auth/backend/backend.go (hunk 1 above)
Comment (savelyev-an) [reply vault-plugins/flant_iam_auth/backend/backend.go:441]: it looks like this method will be run very often, it is just a reminder what to do if in profiling we really  will see the problem with this method

File: vault-plugins/flant_iam_auth/backend/backend.go (hunk 3 above)
Comment (savelyev-an) [reply vault-plugins/flant_iam_auth/backend/backend.go:506]: fixed

File: vault-plugins/flant_iam_auth/backend/path_tenant.go (hunk 4 above)
Comment (savelyev-an) [reply vault-plugins/flant_iam_auth/backend/path_tenant.go:125]: Probably I have done you suggested

**Pull request merged**